dsa_test: Add test for unprivileged applications

ysun opened the pull request:
Intel updated kernel driver to prevent the accelerators from being directly mapped into unprivileged user applications, because malicious may be able to cause corrupt memory.

Add a test when running dsa_test in non-root mode, expecting failed operation.

fenrus75 commented:
wouldn;t it be better to change the tool to just always use the write() method, and outright stop using the mmap approach entirely?

ysun commented:
Thanks @fenrus75 , I get you. Let me refine the patch.

ysun committed:
dsa_test: Use syscall write to submit descriptor

Intel updated kernel driver to prevent the accelerators from being directly
mapped into unprivileged user applications, because malicious may be
able to cause corrupt memory.

Align with the kernel change, write the descriptor to cdev instead of
enqcmd by default when submit shared work queue descriptor. Meanwhile, retain
the enqcmd way if force to use it with new option '-u'.

More details refer to link:
https://www.intel.com/content/www/us/en/security-center/advisory/intel-sa-01084.html

Signed-off-by: Yi Sun <yi.sun@intel.com>

ysun commented:
@fenrus75  @ramesh-thomas Please review

ysun commented:
Hi guys, I get a question here: Does a non-root user have permission to open the cdev /dev/dsa/wq0.0 ? 
Currently, I'm thinking about retaining the use of ENQCMD for root users, while calling write for non-root users.
But a non-root user could normally be blocked at opening the cdev, so have no chance to write the descriptor.

fenrus75 commented:
On Thu, Oct 10, 2024 at 12:49 AM Eason ***@***.***> wrote:

> Hi guys, I get a question here: Does a non-root user have permission to
> open the cdev /dev/dsa/wq0.0 ?
> Currently, I'm thinking about retaining the use of ENQCMD for root users,
> while calling write for non-root users.
> But a non-root user could normally be blocked at opening the cdev, so have
> no chance to write the descriptor.
>

the system administrator can give non-root user permissions for the device
and the device nodes.

however as to your ENQCMD question.... "root users" is a bit of a shortcut.
The actual permission is the CAP_SYS_RAWIO capability, which "root" may or
may not have.
(for example, root in a container won't have it.. nor will root in certain
types of secure boot or other security restricted environments)

how do you plan to check if you have that capability?

ysun commented:
> the system administrator can give non-root user permissions for the device and the device nodes. 
> however as to your ENQCMD question.... "root users" is a bit of a shortcut. The actual permission is the CAP_SYS_RAWIO capability, which "root" may or may not have. (for example, root in a container won't have it.. nor will root in certain types of secure boot or other security restricted environments) how do you plan to check if you have that capability?

For checking the CAP_SYS_RAWIO flag, the first thing that comes to mind is the capget() syscall, or some related library. I'm trying it.
As to the patch, instead of simply replacing enqcmd, my plan is to retain enqcmd when the app has the CAP_SYS_RAWIO capability and switch to use the write func to submit the descriptor otherwise.
I've also noticed that some additional adjustments are needed, such as disabling mmap for non-root users. 
I'll update the PR ASAP. 
Thanks Arjan

fenrus75 commented:
I would strongly suggest always using the write() call.. that way it's more
robust against the conditions of permission for enqcmd changing over time.


On Thu, Oct 10, 2024 at 7:24 AM Eason ***@***.***> wrote:

> the system administrator can give non-root user permissions for the device
> and the device nodes.
> however as to your ENQCMD question.... "root users" is a bit of a
> shortcut. The actual permission is the CAP_SYS_RAWIO capability, which
> "root" may or may not have. (for example, root in a container won't have
> it.. nor will root in certain types of secure boot or other security
> restricted environments) how do you plan to check if you have that
> capability?
>
> For checking the CAP_SYS_RAWIO flag, the first thing that comes to mind is
> the capget() syscall, or some related library. I'm trying it.
> As to the patch, instead of simply replacing enqcmd, my plan is to retain
> enqcmd when the app has the CAP_SYS_RAWIO capability and switch to use the
> write func to submit the descriptor otherwise.
> I've also noticed that some additional adjustments are needed, such as
> disabling mmap for non-root users.
> I'll update the PR ASAP.
> Thanks Arjan
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/intel/idxd-config/pull/73#issuecomment-2405240770>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAJ54FOHXWGWDURJJKKUK5TZ22E2PAVCNFSM6AAAAABOWCFP7SVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDIMBVGI2DANZXGA>
> .
> You are receiving this because you were mentioned.Message ID:
> ***@***.***>
>

ysun commented:
> I would strongly suggest always using the write() call.. that way it's more robust against the conditions of permission for enqcmd changing over time.

Indeed, and that's the simplest approach as well. The patch is implemented like that currently.
However, for the app 'dsa_test', using only write() would miss out on testing the function of enqcmd.

Alternatively, I’d like to add a new option that forces the use of enqcmd without checking the CAP_SYS_RAWIO flag, while keeping write() as the default method.

ysun commented:
Hi @fenrus75  @ramesh-thomas Please have another review.
Add a switch -u toggling ENQCMD and write, which can cover both two approaches as a testing tool.
Take write as default method. I verified the patch on a GNR HW:

|  Permit  | write()  | ENQCMD       |
| -------- | -------- | ------------ |
|  root    | pass     | pass         |
| non-root | pass     | failed when mmap | 

BTW, if the ENQCMD toggle is not accepted, it's easy to remove it and only keep write func.
But from QA perspective, keeping both methods allows us to cover more test cases and code paths, especially since the changes are limited to dsa_test. :)

ysun commented:
@fyu1 Involve Fenghua for review.

ramesh-thomas commented:
Merged to pending branch
